<center>
<h1> TP-Projet d'optimisation numérique </h1>
<h1> Algorithme de Newton </h1>
</center>

## Implémentation 
 
1. Coder l’algorithme de Newton local en respectant la spécification ci-dessous (fichier `Algorithme_De_Newton.jl`)


In [1]:
using LinearAlgebra
using Documenter
using Markdown  
include("Algorithme_De_Newton.jl")
# @doc Algorithme_De_Newton

Algorithme_De_Newton

2. Vérifier que les tests ci-dessous passent.

In [3]:
using Test

# Tolérance pour les tests d'égalité
tol_erreur = sqrt(eps())

## ajouter les fonctions de test
include("../test/fonctions_de_tests.jl")
include("../test/tester_algo_newton.jl")
include("../src/Algorithme_De_Newton.jl")

affiche = false

@testset "Test algo newton" begin
	# Tester l'algorithme de Newton
	tester_algo_newton(affiche,Algorithme_De_Newton)
end;

Test Summary:    | Pass  Total
Test algo newton |   22     22


In [4]:
#using Pkg; Pkg.add("LinearAlgebra"); Pkg.add("Markdown")
# using Documenter
using LinearAlgebra
using Markdown                             # Pour que les docstrings en début des fonctions ne posent
                                           # pas de soucis. Ces docstrings sont utiles pour générer 
                                           # la documentation sous GitHub
include("Algorithme_De_Newton.jl")

# Affichage les sorties de l'algorithme des Régions de confiance
function my_afficher_resultats(algo,nom_fct,point_init,xmin,fxmin,flag,sol_exacte,nbiters)
	println("-------------------------------------------------------------------------")
	printstyled("Résultats de : ",algo, " appliqué à ",nom_fct, " au point initial ", point_init, ":\n",bold=true,color=:blue)
	println("  * xsol = ",xmin)
	println("  * f(xsol) = ",fxmin)
	println("  * nb_iters = ",nbiters)
	println("  * flag = ",flag)
	println("  * sol_exacte : ", sol_exacte)
end

# Fonction f0
# -----------
f0(x) =  sin(x)
# la gradient de la fonction f0
grad_f0(x) = cos(x)
# la hessienne de la fonction f0
hess_f0(x) = -sin(x)
sol_exacte = -pi/2
options = []

x0 = sol_exacte
xmin,f_min,flag,nb_iters = Algorithme_De_Newton(f0,grad_f0,hess_f0,x0,options)
my_afficher_resultats("Newton","f0",x0,xmin,f_min,flag,sol_exacte,nb_iters)
x0 = -pi/2+0.5
xmin,f_min,flag,nb_iters = Algorithme_De_Newton(f0,grad_f0,hess_f0,x0,options)
my_afficher_resultats("Newton","f0",x0,xmin,f_min,flag,sol_exacte,nb_iters)
x0 = pi/2
xmin,f_min,flag,nb_iters = Algorithme_De_Newton(f0,grad_f0,hess_f0,x0,options)
my_afficher_resultats("Newton","f0",x0,xmin,f_min,flag,sol_exacte,nb_iters)

-------------------------------------------------------------------------
Résultats de : Newton appliqué à f0 au point initial -1.5707963267948966:
  * xsol = -1.5707963267948966
  * f(xsol) = -1.0
  * nb_iters = 0
  * flag = 0
  * sol_exacte : -1.5707963267948966
-------------------------------------------------------------------------
Résultats de : Newton appliqué à f0 au point initial -1.0707963267948966:
  * xsol = -1.5707963267949088
  * f(xsol) = -1.0
  * nb_iters = 3
  * flag = 0
  * sol_exacte : -1.5707963267948966
-------------------------------------------------------------------------
Résultats de : Newton appliqué à f0 au point initial 1.5707963267948966:
  * xsol = 1.5707963267948966
  * f(xsol) = 1.0
  * nb_iters = 0
  * flag = 0
  * sol_exacte : -1.5707963267948966


## Interprétation 

1. Justifier les résultats obtenus pour l'exemple $f_0$ ci-dessus;

L'Algorithme de Newton nous permet de retrouver le point pour lequel une fonction atteint son minimum.

f0(x) =  sin(x)

grad_f0(x) = cos(x)

hess_f0(x) = -sin(x)

sol_exacte = -pi/2

Pour f0(x) = sin(x), la valeur minimun de cette fonction est bien -1. On obtient bien cette valeur pour x = k*pi/2 pour tout k entier relatif.

Cas 1:
On donne comme poit initial x0 = -pi/2, qui est solution du problème. Les conditions de terminaison sont donc vérifiées et on entre donc pas dans la boucle.

Cas 2: 
On donne comme poit initial x0 = -pi/2, qui est n'est pas extremun de f0. 
On calcule à chaque itération, la direction de Newton d : $$\nabla^2 f(x) d = - \nabla f(x)$$
Sachant que $$\nabla^2 f0(x) = - \sin(x)$$ et $$\nabla^2 f0(x) = \cos(x)$$ 
on obtient $$ d = 1 / \tan(x) $$

A la première itération, on obtient d = -0.54630248984 et x = -1.61709881663 
A la deuxième itération, on obtient d = 0.04633560785 et x = -1.57076320878 
A la troisième itération, on obtient d = -0.00003311801 et x = -1.57079632679
La solution exacte étant -1.5707963267948966, on résout le problème en 3 itérations.

Cas 3:
On donne comme poit initial x0 = pi/2. Cas similaire au cas 1.

2. Soit 
$$ f_{1} : \mathbb{R}^3 \rightarrow \mathbb{R}$$ $$ (x_1,x_2, x_3) \mapsto  2 (x_1 +x_2 + x_3 -3)^2 + (x_1-x_2)^2 + (x_2 - x_3)^2$$ 

Justifier que l’algorithme implémenté converge en une itération pour $f_{1}$;
$$ f_{1}$$ peut s'écrire sous la forme d'une fonction quadratique.
$$ f(x) = 1/2 x^T A x - b^T x + c $$
et on a :
$$ \nabla f(x) = A x - b $$
$$ \nabla^2 f(x) = A $$ 
Sachant que l'on calcule, à chaque itération, $$d= x_{k+1} - x_{k}  = - \nabla^2 f^(-1)(x) \nabla f(x))$$
En remplaçant, on obtient : $$ x_{k+1} = A^(-1) b $$
Ensuite les conditions de terminaison sont directement vérifiées, car en remplaçant avec $$ x_{k+1} $$, on obtient un gradient nulle.

3. Soit 
$$ f_{2} : \mathbb{R}^2 \rightarrow \mathbb{R}$$ $$ (x_1,x_2) \mapsto 100(x_2-x_1^2)^2 + (1-x_1)^2 $$ 

Justifier que l’algorithme puisse ne pas converger pour $f_{2}$ avec certains points initiaux.

L'algori

